modules/client-handler/src/main/java/org/apache/ignite/client/handler/ClientHandlerMetricSource.java

```diff
@@ -0,0 +1,418 @@
+        }
+    }
+
+    /**
+     * Decrements active transactions.
+     */
+    public void transactionsActiveDecrement() {
+        Holder h = holder;
+
+        if (h != null) {
+            h.transactionsActive.decrement();
+        }
+    }
+
+    /**
+     * Gets active cursors.
+     *
+     * @return Active cursors.
+     */
+    public long cursorsActive() {
+        Holder h = holder;
+
+        return h == null ? 0 : h.cursorsActive.value();
+    }
+
+    /**
+     * Increments active cursors.
+     */
+    public void cursorsActiveIncrement() {
+        Holder h = holder;
+
+        if (h != null) {
+            h.cursorsActive.increment();
+        }
+    }
+
+    /**
+     * Decrements active cursors.
+     */
+    public void cursorsActiveDecrement() {
+        Holder h = holder;
+
+        if (h != null) {
+            h.cursorsActive.decrement();
+        }
+    }
+
+    @Override
+    public synchronized @Nullable MetricSet enable() {
+        Holder h = holder;
+
+        if (h == null) {
+            h = new Holder();
+            holder = h;
+        }
+
+        return h.metricSet;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public synchronized void disable() {
+        holder = null;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public synchronized boolean enabled() {
+        return holder != null;
+    }
+
+    private static class Holder {
+        private final AtomicLongMetric connectionsInitiated =
+                new AtomicLongMetric("connections.Initiated", "Total initiated connections");
+
+        private final AtomicLongMetric sessionsAccepted =
+                new AtomicLongMetric("sessions.Accepted", "Total accepted sessions");
+
+        private final AtomicLongMetric sessionsActive =
+                new AtomicLongMetric("sessions.Active", "Active sessions");
+
+        private final AtomicLongMetric sessionsRejected =
+                new AtomicLongMetric("sessions.Rejected", "Total sessions rejected due to handshake errors");
+
+        private final AtomicLongMetric sessionsRejectedTls =
+                new AtomicLongMetric("sessions.rejected.Tls", "Total sessions rejected due to TLS handshake errors");
+
+        private final AtomicLongMetric sessionsRejectedTimeout =
+                new AtomicLongMetric("sessions.rejected.Timeout", "Total sessions rejected by timeout");
+
+        private final AtomicLongMetric bytesSent = new AtomicLongMetric("bytes.Sent", "Total bytes sent");
+
+        private final AtomicLongMetric bytesReceived = new AtomicLongMetric("bytes.Received", "Total bytes received");
+
+        private final AtomicLongMetric requestsActive = new AtomicLongMetric("requests.Active", "Requests in progress");
+
+        private final AtomicLongMetric requestsProcessed = new AtomicLongMetric("requests.Processed", "Total processed requests");
+
+        private final AtomicLongMetric requestsFailed = new AtomicLongMetric("requests.Failed", "Total failed requests");
```

Comment:
I wonder what is the performance impact of metrics like this? Is it OK to use them like this?

Reply:
Can you please elaborate? What is your concern?

Reply:
That's OK, I think I had some wrong idea here, sorry.